### PR TITLE
iSCSI support

### PR DESCRIPTION
**Pull request by mvollmer**
Depends on #3680.

### CODE REVIEW

**Comment by mvollmer** [comment]
The call to GetInitiatorName in iscsi_discover is unnecessary.

**Comment by mvollmer** [comment]
> The call to GetInitiatorName in iscsi_discover is unnecessary.

Fixed.

**Comment by mvollmer** [comment]
Added authentication and tested with a local storaged master build.

**Comment by mvollmer** [comment]
Rebased and modified to fit the new "+" button design of the overview panels.

**Comment by andreasn** [comment]
updated mockup with renaming for initiator https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/storage/iscsi-v2.png

**Comment by stefwalter** [comment]
@mvollmer @andreasn Does the default ISCSI initiator name work out of the box? Because if you have to set the name before proceeding, then it's very undiscoverable.

**Comment by andreasn** [comment]
It would work, as long as you've set up the iSCSI Target server auth to pick up on that name.
I think the default is usually something like 1994-03.redhat.com:client so in a lot of cases it's "wrong".
One thing I had in a earlier version of the design was a box to change the name of the initiator in the join dialog, but that somewhat indicates that you can change it on the fly for like the 3rd target you're adding (and then destroying the connection for the first two) and that it's a per target kind of setting, that it isn't.

So I wonder if we should be clearer about the initiator name, either in the panel or in the add dialog somewhere. I don't want changing the name to easy, since it's potentially a quite destructive action.

**Comment by andreasn** [comment]
Actually, on my system it's iqn.1994-05.com.redhat:7d366003913
So probably not a excellent default.

**Comment by mvollmer** [comment]
> Does the default ISCSI initiator name work out of the box?

I have no hands-on experience with this, unfortunately.  My understanding is that for a completely new box, the initial default initiator name works out of the box.  If the box is later reinstalled from scratch but should keep its identity (so that existing ACLs in targets keep working as expected), then one needs to change the initiator name back to its old value.

So, we were working with the assumption that it is rare to change the initiator name, but when you need to do it, it is important that you can.

**Comment by stefwalter** [comment]
> I have no hands-on experience with this, unfortunately. My understanding is that for a completely new box, the initial default initiator name works out of the box. If the box is later reinstalled from scratch but should keep its identity (so that existing ACLs in targets keep working as expected), then one needs to change the initiator name back to its old value.

Okay, then it seems okay if this is a secondary option, at least for now.

**Comment by stefwalter** [comment]
In general we should only be using the quick delete [check] option when there are:

 * Many things to delete.
 * It should be possible to delete them rapidly.

I don't think that applies in this case. @andreasn WDYT?

**Comment by andreasn** [comment]
Don't we also use them as a saftey-padding for destructive actions? Avoiding accidental clicks and such.
In this case, since storage is involved, "deleting" (rather logging out from) a iscsi target could really mess up your system, I didn't want it as easy accessible.
Should we do a warning dialog instead? Feels less elegant.

**Comment by andreasn** [comment]
In order to cope with the discoverability of the initiator name, we could set the add dialogs header to "Add iSCSI portal to name-of-my-initiator"

**Comment by mvollmer** [comment]
Discovery needs to be cancellable.

**Comment by fabiand** [comment]
I really like the flow and how dynamic the whole thing is.

A few notes:

- The discovery dialog should accept <return> to continue / go to the
target dialog
- Allow to select multiple targets
- Display errors if the occur, i.e. tried to connect to a portal, got the list of targets, selected one, and then - after pressing add - it failed silently. Retrying this with the same target I then got an error that the session already exists.

In the end I don't know what was wrong with the target I used.

**Comment by fabiand** [comment]
Another thought - It would be good if the iscsi disks (or the bus of a disk in general) could be somehow shown/reflected in the disk list.

**Comment by andreasn** [comment]
> Allow to select multiple targets

Maybe.
I think this is fixable with checkboxes or so. One worry though, can different targets have different credentials?
We would need to throw up different credentials screens for every target if so. Gets messy quite quick.

**Comment by fabiand** [comment]
> One worry though, can different targets have different credentials?
> We would need to throw up different credentials screens for every target if so. Gets messy quite quick.

Yes, good point. You can have target specific ACLs.

**Comment by andreasn** [comment]
Odd looking styling:
screenshot from 2015-10-05 12-26-18

**Comment by andreasn** [comment]
screenshot from 2015-10-05 12-27-43

Needs text-overflow: ellipsis;

**Comment by andreasn** [comment]
Error is misaligned on the left side. Doesn't line up with the box above.
screenshot from 2015-10-05 12-42-20

**Comment by andreasn** [comment]
@fabiand explained that multi-select would be useful for multipath devices (as they will show up as 2 devices now), but that ideally the system should be able to figure out that it is the same disk and consolidate them somehow. I think that sounds like the best solution.

**Comment by fabiand** [comment]
Multipath targets can be identified through the serial of the disk: All
disks with the same serial will be a single disk with multiple paths.

On Mon, Oct 5, 2015 at 12:45 PM, andreasn <notifications@github.com> wrote:

> @fabiand <https://github.com/fabiand> explained that multi-select would
> be useful for multipath devices (as they will show up as 2 devices now),
> but that ideally the system should be able to figure out that it is the
> same disk and consolidate them somehow. I think that sounds like the best
> solution.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockpit-project/cockpit/pull/2807#issuecomment-145492882>
> .
>



-- 
Fabian Deutsch <fdeutsch@redhat.com>
RHEV Hypervisor
Red Hat

**Comment by andreasn** [comment]
Probably on the todo already, but I tried to rename a initiator to "Hello?", and unfortunately succeeded.

**Comment by mvollmer** [comment]
> Probably on the todo already, but I tried to rename a initiator to "Hello?", and unfortunately succeeded.

It is probably technically possible to use any string, actually, as long as the target can be configured accordingly.  I tried "Hello?" and could successfully connect to a iSCSI target that didn't require authentication.

Storaged just allows any non-empty string as the initiator name.  I have filed https://github.com/storaged-project/storaged/issues/32 to maybe improve that.

**Comment by andreasn** [comment]
> It is probably technically possible to use any string, actually, as long as the target can be configured accordingly. I tried "Hello?" and could successfully connect to a iSCSI target that didn't require authentication.

Ah. I see.
Well, if the spec says those names are OK, I guess they are.

**Comment by mvollmer** [comment]
> Well, if the spec says those names are OK, I guess they are.

There are some rules, and we should probably warn people when they break them...  But there are a million input validations that we don't do, actually, and I would really like to get some help from the lower layers of the storage stack for this.

**Comment by mvollmer** [comment]
> Discovery needs to be cancellable.

Done.

**Comment by mvollmer** [comment]
@fabiand

> The discovery dialog should accept to continue / go to the target dialog

I don't understand this.  Could you elaborate?

> Allow to select multiple targets

This will come later with separate mock ups / pull request.

> Display errors if the occur, i.e. tried to connect to a portal, got the list of targets, selected one, and then - after pressing add - it failed silently. Retrying this with the same target I then got an error that the session already exists.

I amnot sure I understand this either... We do display errors as we get them from storaged.  Your description sounds as if the login actually succeeded the first time around.

**Comment by mvollmer** [comment]
> Odd looking styling:

Ouch, the table header was selectable... fixed.

**Comment by mvollmer** [comment]
> Needs text-overflow: ellipsis;

I did that, but it didn't have any effect.  Could you try your CSS magic?

**Comment by mvollmer** [comment]
> It would be good if the iscsi disks (or the bus of a disk in general) could be somehow shown/reflected in the disk list.

It is, isn't it?

**Comment by fabiand** [comment]
>     The discovery dialog should accept to continue / go to the target dialog
> 
> I don't understand this. Could you elaborate?

Sorry, I can't recall what I meant here.

>     Display errors if the occur, i.e. tried to connect to a portal, got the list of targets, selected one, and then - after pressing add - it failed silently. Retrying this with the same target I then got an error that the session already exists.
> 
> I am not sure I understand this either... We do display errors as we get them from storaged. Your description sounds as if the login actually succeeded the first time around.

The problem was that some error appeared, but no error was shown in the UI.
The login seems to have succeeded but the disk did not appear in the list of disks.
The problem was probably that storaged failed to attach the disk, but failed to pass this error to cockpit, so cockpit had no chance to show an error.

**Comment by andreasn** [comment]
> I did that, but it didn't have any effect. Could you try your CSS magic?

Yup, I'll add a fix.

**Comment by andreasn** [comment]
> Yup, I'll add a fix.

I realized this is fixed for all storage objects in master already.

**Comment by mvollmer** [comment]
This is unblocked on fedora-testing (Fedora 23 plus updates-testing).

**Comment by mvollmer** [comment]
For the record, check-storage-iscsi has passed on fedora-atomic: https://github.com/mvollmer/cockpit/commit/e75c0b3f0968010d5f117e53505cec90465ff876

**Comment by dperpeet** [comment]
needs rebase

**Comment by mvollmer** [comment]
> needs rebase:

Done.

File: pkg/storaged/index.html [hunk 1]
```diff
@@ -134,6 +134,44 @@ <h1 translatable="yes">The "storaged" API is not available on this system.</h1>
...
+      <table class="table">
+        {{#Sessions}}
+        <tr>
+          <td style="width: 48px">
+            <div><img  src="images/storage-array.png"></div>
+          </td>
+          <td class="row">
+            <span class="col-md-12">{{Name}}</span>
+            <br>
+            <span class="col-md-12 col-lg-5" translatable="yes">{{Address}}:{{Port}}</span>
```
Comment (stefwalter) [inline pkg/storaged/index.html]: This doesn't need to be translatable.

File: pkg/storaged/dialog.js [hunk 2]
```diff
@@ -122,19 +155,41 @@ define([
...
+                                err = def.Action.failure_filter(vals, err);
+                            if (err) {
+                                if (err.length)
+                                    err = err.map(error_field_to_target);
+                                else
+                                    err = error_field_to_target(err);
+                                $dialog.dialog('failure', err);
+                            }
+                        });
+                } else
```
Comment (stefwalter) [inline pkg/storaged/dialog.js]: I would suggest using braces after, if braces are used before it.

File: pkg/storaged/dialog.js [hunk 3]
```diff
@@ -122,19 +155,41 @@ define([
...
         $dialog.on('change input', function () {
-            clear_errors();
             update_visibility();
         });
 
-        $dialog.on('hidden.bs.modal', function () {
-            $dialog.remove();
-        });
+        function error_field_to_target(err) {
+            if (err.field)
```
Comment (stefwalter) [inline pkg/storaged/dialog.js]: I would suggest using braces surrounding a multi-line conditional statement.

File: pkg/storaged/dialog.js [hunk 4]
```diff
@@ -79,7 +106,13 @@ define([
                         if (e.checked)
                             vals[n].push(f.Options[i].value);
                     });
-                }
+                } else if (f.SelectRow) {
+                    $f.find('tbody tr').each(function (i, e) {
+                        if ($(e).hasClass('highlight'))
+                            vals[n] = f.Rows[i].value;
+                    });
+                } else if (f.Arrow)
```
Comment (stefwalter) [inline pkg/storaged/dialog.js]: Lets use braces around the else if portion too.

File: pkg/storaged/dialog.js [hunk 5]
```diff
@@ -33,28 +33,55 @@ define([
...
+
+        if (cur_dialog)
+            cur_dialog.modal('hide');
+
         var $dialog = $(mustache.render(storage_dialog_tmpl, def));
         $('body').append($dialog);
+        cur_dialog = $dialog;
+
+        $dialog.on('hidden.bs.modal', function () {
+            $dialog.remove();
```
Comment (stefwalter) [inline pkg/storaged/dialog.js]: Should this clear cur_dialog if it's === to $dialog?

**Pull request closed without merging**